Add RTC driver for STM32L4 using LL APIs

kzyapkov opened the pull request:
This reworks the RTC driver to use LL APIs instead of HAL. Only tested on L4xx.

Commit:
drivers: rtc: add support for STM32 RTC

Add support for the STM32 Real-Time-Clock leveraging the LL APIs

Signed-off-by: Johannes Hutter <johannes@proglove.de>
Signed-off-by: Kiril Zyapkov <k.zyapkov@allterco.com>

Olivier-ProGlove committed:
dts/rtc: Introduce binding for STM32 RTC

This commit adds rtc fixup and dts blocks for RTC hardware
on STM32L4.

Signed-off-by: Olivier Martin <olivier.martin@proglove.de>

Olivier-ProGlove commented:
I was actually reworking my pull-request. But I am happy to help you on your pull-request  if you have more time to focus on it than I have (not much!).

I was planning to make sure the STM32 RTC driver pass the Zephyr test `tests/drivers/rtc/rtc_basic_api`. At the moment it does not build. I noticed you made similar changes to fix the build (in your case the RTC sample).

I have actually two Nucleo STM32 boards here: `L476RG` and `F411RE`. Would you mind to make the needed changes to make these board build. So I can help you to test your pull-request

Changes to be made, in the board DTS:
```
&rtc {
       status = "ok";
};
```

and in the board `defconfig`:

```
# Enable RTC
CONFIG_RTC_STM32=y
```

kzyapkov commented:
@Olivier-ProGlove I am tempted to enable the symbol for the entire series, but not sure if this is the right approach, and would it require `status = "ok";` in dts files as well.

erwango commented:
@kzyapkov , @JoeHut, @Olivier-ProGlove: RTC driver is nice and having LL version is even nicer, but can we squash all of that and introduce directly the LL version, so we don't review code that disappear in the next commit.

JoeHut commented:
@erwango Sure, I'll take care of synchronizing us three on that. Sorry for the mess.

erwango commented:
@JoeHut , np! Thanks to all 3 of you for this work.

Olivier-ProGlove commented:
I will actually stop the work I was doing on my pull-request (and close it to avoid confusion for new comers). I am guessing @kzyapkov took in account the comments that were made in my PR.

@kzyapkov feel free to squash the commits.

Commit:
boards/arm: Enable RTC on STM32L4 boards

All supported boards from the STM32L4 series have RTC enabled
in dts.

Signed-off-by: Kiril Zyapkov <k.zyapkov@allterco.com>

kzyapkov commented:
Squashed as requested, with a number of fixes. Still needs sub-second precision for alarms and STM32F4 support.

JoeHut commented:
I just tested the sample on nucleo_l476rg and it works fine :+1:

kzyapkov commented:
As discussed with @JoeHut, removed F4 references and changed the Signed-Off-By header of the problematic commit to match the author email.

ydamigos commented on dts/arm/st/stm32l4.dtsi:
```diff
@@ -285,6 +285,16 @@
 			phase_seg2 = <6>;
 		};
+
+		rtc: rtc@40002800 {
```
The rtc node property `prescaler` is missing. It is defined as required in rtc.yaml.

kzyapkov replied:
Not sure what its semantics would be for this driver.

ydamigos replied:
If it is not necessary for all SoCs, we should make it optional in rtc.yaml.

ydamigos commented on include/rtc.h:
```diff
@@ -15,7 +15,11 @@ extern "C" {
 #endif
 
 /** Number of RTC ticks in a second */
+#ifdef CONFIG_RTC_PRESCALER
```
I couldn't find RTC_PRESCALER option in Kconfig files.

kzyapkov replied:
It is only defined by RTC driver implementations and not really applicable for this hardware. There is no direct access to the RTC counter on STM32, so the driver operates in milliseconds and we directly define RTC_ALARM_SECOND to be used as time base.

ydamigos replied:
The description of RTC_ALARM_SECOND is the same as the prescaler property in DT:
`RTC prescaler used to determine ticks per second`. It is confusing.

kzyapkov replied:
I will revert the change in `rtc.h`, but please see below.

kzyapkov replied:
Actually, `#define RTC_ALARM_SECOND (32768 / CONFIG_RTC_PRESCALER)` and we need RTC_ALARM_SECOND to equal 1000, so it's not the same. CONFIG_RTC_PRESCALER would need to equal 32.768

ydamigos replied:
I wasn't clear. 
```
config RTC_ALARM_SECOND
	int
	default 1000
	prompt "Number of RTC ticks per second"
	help
	  RTC prescaler used to determine ticks per second
```
The prompt and help section of `RTC_ALARM_SECOND` option should be updated. They are confusing because they match the description of the prescaler property in rtc.yaml.
```
    prescaler:
      type: int
      category: required
      description: Prescaler value denotes number of RTC ticks in 1 second
      generation: define
```

ydamigos commented on drivers/rtc/Kconfig.stm32_rtc:
```diff
@@ -0,0 +1,29 @@
+
+if RTC_STM32
+
+config RTC_ALARM_SECOND
```
Could you use rtc node `prescaler` property from DT instead of adding a Kconfig option?

kzyapkov replied:
Added, along with a line in `arm/soc/st_stm32/stm32l4/dts.fixup`

    #define CONFIG_RTC_0_PRESCALER                  ST_STM32_RTC_40002800_PRESCALER

However, `rtc.h` expects `CONFIG_RTC_PRESCALER`, not sure where this mapping should go

ydamigos replied:
If there is only one RTC then you can just use CONFIG_RTC_*, like we did for USB interface.

ydamigos reviewed (changes requested):
Please use prescaler property from DT to describe the number of RTC ticks per second instead of adding a Kconfig option.

ydamigos reviewed (changes requested):
Some more comments:

Why `st,stm32-rtc-v1.yaml` and not just `st,stm32-rtc.yaml`?
You missed to enable rtc on disco_l475_iot1. I tested rtc on the board and it worked.

kzyapkov commented:
@ydamigos thanks for going over this. I am still not sure how to handle the prescaler issue. Are we changing the meaning of `CONFIG_RTC_PRESCALER`? How should then RTC_ALARM_SECOND be calculated? Should this config be sourced from dts or Kconfig?

ydamigos commented:
@kzyapkov RTC_ALARM_SECOND in RTC api is the number of RTC ticks per second. In STM32 RTC driver it is defined as 1000 (default value). As a result, I understand that, in STM32 case, 1 second is 1000 RTC ticks (resolution 1 ms). Is it correct?

The RTC api defines `u32_t alarm_val` as 32bit RTC alarm value. In STM32 RTC driver, it is handled as milliseconds. Is this correct? Do the other RTC implementations handle it as milliseconds?

>Maybe we should use both clock-frequency and prescaler from DTS to set the time base?

Yes, we should get the clock-frequency and prescaler (if we need a prescaler value for STM32 RTC) from DT.

There is an implementation of stm32f4 in linux kernel. Did you check it? Maybe, the DT bindings in linux kernel give us an idea how to proceed.

kzyapkov commented:
> in STM32 case, 1 second is 1000 RTC ticks (resolution 1 ms). Is it correct?

Yes, but the unit is "arbitrarily" chosen as time base, this is not what the hardware uses. Alas, HW does not have an ordinary counter.

> Do the other RTC implementations handle it as milliseconds?

No. Other implementations define `CONFIG_RTC_PRESCALER` (QMSI in Kconfig; MCUX in DT) which is then used to scale RTC_ALARM_SECOND, assuming a 32.768KHz clock.

Looking at the RTC driver subsystem in linux right now which seems to expose a richer API. Can you point to the stm32f4 implementation?

ydamigos commented:
> Can you point to the stm32f4 implementation?

STM32 RTC linux driver:
https://elixir.bootlin.com/linux/latest/source/drivers/rtc/rtc-stm32.c

DT bindings:
https://elixir.bootlin.com/linux/v4.17.2/source/Documentation/devicetree/bindings/rtc/st,stm32-rtc.txt

stm32f429.dtsi:
https://elixir.bootlin.com/linux/v4.17.2/source/arch/arm/boot/dts/stm32f429.dtsi

stm32f746.dtsi:
https://elixir.bootlin.com/linux/v4.17.2/source/arch/arm/boot/dts/stm32f746.dtsi

kzyapkov commented:
So the Linux RTC API operates with calendar time natively. Unless Zephyr chooses to expand the RTC API with calendar time, this is my plan:
 * internally, operate with a 64 bit counter which would be the UNIX time stamp in milliseconds or native RTC resolution (RTC_STM32_PRESCALER_S)
 * implement proper Gregorian Date to UNIX timestamp conversion
 * use the least-significant 32 bits of the 64-bit counter for the public read, config and alarm APIs.

Any guidance or feedback is welcome.

Olivier-ProGlove commented:
> Why `st,stm32-rtc-v1.yaml` and not just `st,stm32-rtc.yaml`?

@ydamigos I based the name of the new file on an existing file. Looking at the other RTC `.yaml`, yes there is no reason to add `-v1`. The yaml file can be renamed.

kzyapkov commented:
@ydamigos @erwango @JoeHut @Olivier-ProGlove,

So I did another round of rework, addressed most comments. Important points:

 * the driver now selects NEWLIB_LIBC and uses `mktime` and `gmtime_r` to convert to and from a monotonic timer.
 * dropped sub-second resolution, so values read and set for alarms have a resolution of 1 s. In sync with hardware capabilities, the values are "number of seconds since 2000-01-01 00:00:00". Hardware supports years between 2000 and 2099.
 * DT is used to set the prescaler so that `RTC_ALARM_SECOND == 1`

The driver uses Cube's LL layer for EXTI. Not sure if this is the correct approach in this case, or if `exti_stm32` should be used instead.

kzyapkov commented:
Also, not sure why tests are failing.

erwango commented:
@kzyapkov , one part of CI issues is due to https://github.com/zephyrproject-rtos/zephyr/issues/8593
For the other part with "log too large", I don't know.

kzyapkov commented:
I'm actually working on that, will update the PR within a couple of days when I get a spare minute. I intend to only support LSI and LSE.

kzyapkov commented:
Added LSI/LSE clock source selection, rebased.

Is this affected by #8340 ?

erwango commented on drivers/rtc/Kconfig.stm32_rtc:
```diff
@@ -0,0 +1,70 @@
+config RTC_STM32
+	bool "STM32 RTC Driver"
+	depends on SOC_FAMILY_STM32 && SOC_SERIES_STM32L4X
+	default n
```
Remove 'default n'

erwango commented on drivers/rtc/Kconfig.stm32_rtc:
```diff
@@ -0,0 +1,70 @@
+
+config RTC_STM32
+	bool "STM32 RTC Driver"
+	depends on SOC_FAMILY_STM32 && SOC_SERIES_STM32L4X
```
Remove dependency on L4. It is sufficient to control activation per series (in arch)

erwango commented on drivers/rtc/Kconfig.stm32_rtc:
```diff
@@ -0,0 +1,70 @@
+	help
+	  Build STM32 RTC driver.
+
+choice
```
For easier use, add a name to the choice and remove default.
Default value will be the first option listed

erwango commented on drivers/rtc/Kconfig.stm32_rtc:
```diff
@@ -0,0 +1,70 @@
+choice
+	bool "RTC clock source"
+	depends on RTC_STM32
+	default RTC_STM32_CLOCK_SRC_LSI
```
remove default

erwango commented on drivers/rtc/Kconfig.stm32_rtc:
```diff
@@ -0,0 +1,70 @@
+
+endchoice #RTC_STM32_CLOCK_SRC
+
+choice
```
Same here, add a name to the choice

erwango commented on drivers/rtc/Kconfig.stm32_rtc:
```diff
@@ -0,0 +1,70 @@
+choice
+	prompt "LSE oscillator drive capability"
+	depends on RTC_STM32_CLOCK_SRC_LSE
+	default RTC_STM32_CLOCK_SRC_LSE_DRIVE_LOW
```
Remove default

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -0,0 +1,360 @@
+#define RTC_STM32_PRESCALER_A 0x7F
+#define RTC_STM32_PRESCALER_S 0x00F9
+
+#else
```
#else /* CONFIG_RTC_STM32_CLOCK_SRC_LSE */

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -0,0 +1,360 @@
+#ifdef CONFIG_SYS_POWER_DEEP_SLEEP
+static int rtc_suspend_device(struct device *dev)
+{
+	rtc_stm32_disable(dev);
```
You should not disable RTC when entering Sleep mode, it will halt the RTC.
Low power should actually be handlded by the PWR IP which should map OS power management requests to SoC power states.
RTC behavior (and for other IP it is the same) will be altered depending on PWR state that will be defined (CF "RTC and low-power modes" in reference manual.
This being said, I think you should remove power management from this driver.

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -0,0 +1,360 @@
+
+static int rtc_stm32_init(struct device *dev)
+{
+	static const struct stm32_pclken pclken = {
```
Add this in driver _config struct (cf other drivers like I2C, SPI, ...)

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -0,0 +1,360 @@
+		.bus = STM32_CLOCK_BUS_APB1,
+	};
+
+	LL_RTC_InitTypeDef rtc_config = {
```
Same here, add this in driver config struct

kzyapkov replied:
The LL API won't take `const` here, but I'll try

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -0,0 +1,360 @@
+	LL_RTC_EnableShadowRegBypass(RTC);
+
+	/* Install RTC interrupt */
+	IRQ_CONNECT(CONFIG_RTC_0_IRQ, CONFIG_RTC_0_IRQ_PRI,
```
Move this to driver irq function.

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -0,0 +1,360 @@
+		.get_pending_int = rtc_stm32_get_pending_int,
+};
+
+DEVICE_DEFINE(rtc_stm32, CONFIG_RTC_0_NAME,
```
Please use DEVICE_AND_API_INIT instead

erwango reviewed (changes requested):
Some comments.
It seems you based you driver structure on a "_qmsi" example.
Unfortunately these are not the most up to date structure and it would be better to use other reference like stm32 i2c or spi

kzyapkov commented:
Tried to address all comments.

kzyapkov commented:
Unlike the Linux kernel, DT here doesn't include the complete clock config and other HW settings

kzyapkov reviewed (commented):
Unlike the linux kernel, DT doesn't specify the complete clock config and other hw peculiarities here.

erwango commented on arch/arm/soc/st_stm32/stm32l4/dts.fixup:
```diff
@@ -62,6 +62,12 @@
+#define CONFIG_RTC_0_IRQ_PRI                    ST_STM32_RTC_40002800_IRQ_0_PRIORITY
+#define CONFIG_RTC_0_IRQ                        ST_STM32_RTC_40002800_IRQ_0
+#define CONFIG_RTC_0_NAME                       ST_STM32_RTC_40002800_LABEL
+#define CONFIG_RTC_PRESCALER                    ST_STM32_RTC_40002800_PRESCALER
```
Where is that used?

kzyapkov replied:
Only `...BASE_ADDRESS` isn't used anywhere, just there for completeness. The address is used as defined by the ll drivers, simply `RTC`. `CONFIG_RTC_PRESCALER` is used in `include/rtc.h` to scale `RTC_ALARM_SECOND`.

The rest are used by the `INIT` macro and `irq_config()`.

erwango replied:
Ok. Not sure it is the right approach, but since used already in NXP driver, let's keep it that way.

kzyapkov commented on drivers/rtc/Kconfig.stm32_rtc:
```diff
@@ -0,0 +1,67 @@
+	select USE_STM32_LL_EXTI
+	select NEWLIB_LIBC
+	help
+	  Build STM32 RTC driver.
```
done

erwango commented:
@kzyapkov , approved, but please check with @ydamigos if he's satisfied

ydamigos reviewed (approved):
I tested it on STM32F4 family (96b_carbon) and it works.
@kzyapkov I opened a PR to your github tree with the patches for STM32F4 series, if you want to merge them with this PR. If not, I will create a new PR when this is merged.

kzyapkov commented:
@ydamigos saw that too late. I did try to test it on a Nucleo F401 (the only F4 I own), but spent a day or so trying to figure out why it won't boot. Turns out my ancient board has a different default clock setup.

Thanks all for the feedback!
